blockchain/chain_sync/src/bad_block_cache.rs

```diff
@@ -0,0 +1,43 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use async_std::sync::RwLock;
+use cid::Cid;
+use lru::LruCache;
+
+/// Threadsafe cache for tracking bad blocks
+#[derive(Debug)]
+pub struct BadBlockCache {
+    cache: RwLock<LruCache<Cid, String>>,
+}
+
+impl Default for BadBlockCache {
+    fn default() -> Self {
+        Self::new(1 << 15)
+    }
+}
+
+impl BadBlockCache {
+    pub fn new(cap: usize) -> Self {
+        Self {
+            cache: RwLock::new(LruCache::new(cap)),
+        }
+    }
+
+    /// Puts a bad block Cid in the cache with a given reason.
+    pub async fn put(&self, c: Cid, reason: String) -> Option<String> {
```

Comment:
Is there a reason why you wanted to make the reason a String? I would've done a enum similar to the Error Types we have for the Actors

Reply:
for what benefit? The Error in actors isn't itself an enum btw, but I get the point

Reply:
there just isn't a reason to distinguish specifically yet, and the flexibility is nice. Is a good idea though and I may look into it before this PR comes in

Reply:
I made the change, but decided to force push back because using an enum for this uses 56 bytes rather than the 24 needed for a string. There is no functional gain for this, so the extra bytes on such a large cache is counterproductive